### ISSUE TITLE
ClickHouse failing to start when scanning part files

### ISSUE THREAD

**Issue by dylanzr** [yandex/ClickHouse#2417]
## Expected Behavior
Old part files should be garbage collected and should not prevent proper server startup.

## Current Behavior
One of our ClickHouse servers has higher than expected disk space utilization. There are over 1 million part files on disk, and it is taking around 14 hours to scan them before starting tables. This is bad in itself, but the last 2 startups have resulted in an error that appears to be related to trying to access a tmp path that has been deleted.

## Steps to Reproduce
1. Restart clickhouse-server with millions of part files.

## Context (Environment)
System:
CPU cores: 56
RAM: 128G
OS: CentOS 7.4
Disk: 7x4T HDD, RAID0
ClickHouse Build: 1.1.54382

## Detailed Description
Restarts result in the following:
```
2018.05.23 18:45:29.404159 [ 2 ] <Debug> db.logs_shard (Data): Loading data parts
2018.05.24 08:19:15.650093 [ 2 ] <Debug> db.logs_shard (Data): Loaded data parts (1025792 items)
2018.05.24 08:19:15.811672 [ 1 ] <Information> DatabaseOrdinary (db): Starting up tables.
2018.05.24 08:19:17.363693 [ 18 ] <Trace> db.logs_shard (Data): Found 1023853 old parts to remove.
...
lots of lines later
...
2018.05.24 08:21:13.481998 [ 18 ] <Debug> db.logs_shard (Data): Removing part from filesystem 032e3e9464547a440f0194235c9eeb94_4268213_4268213_0
2018.05.24 08:21:14.194246 [ 12 ] <Warning> db.logs_shard (Data): Removing temporary directory /data2/clickhouse/data/db/logs_shard/tmp_delete_032e3e9464547a440f0194235c9eeb94_4268213_4268213_0
2018.05.24 08:21:16.751362 [ 9 ] <Trace> ColumnGathererStream: Gathered column sc_bytes (8.00 bytes/elem.) in 4.66 sec., 1.41338e+07 rows/sec., 107.833 MiB/sec.
2018.05.24 08:21:17.534785 [ 9 ] <Debug> MemoryTracker: Peak memory usage: 76.81 MiB.
2018.05.24 08:21:17.534824 [ 9 ] <Information> db.logs_shard (StorageMergeTree): Cancelled merging parts
2018.05.24 08:21:30.054975 [ 7 ] <Trace> ColumnGathererStream: Gathered column c_ip (22.11 bytes/elem.) in 44.03 sec., 6.12877e+06 rows/sec., 129.223 MiB/sec.
2018.05.24 08:21:30.688387 [ 7 ] <Debug> MemoryTracker: Peak memory usage: 39.71 MiB.
2018.05.24 08:21:30.688437 [ 7 ] <Information> db.logs_shard (StorageMergeTree): Cancelled merging parts
2018.05.24 08:21:42.030808 [ 8 ] <Debug> MemoryTracker: Peak memory usage: 90.55 MiB.
2018.05.24 08:21:42.030887 [ 5 ] <Debug> MemoryTracker: Peak memory usage: 180.11 MiB.
2018.05.24 08:21:42.031035 [ 17 ] <Debug> MemoryTracker: Peak memory usage: 77.59 MiB.
2018.05.24 08:21:42.031083 [ 10 ] <Debug> MemoryTracker: Peak memory usage: 69.80 MiB.
2018.05.24 08:21:42.031218 [ 13 ] <Debug> MemoryTracker: Peak memory usage: 77.56 MiB.
2018.05.24 08:21:42.031247 [ 12 ] <Debug> MemoryTracker: Peak memory usage: 56.00 MiB.
2018.05.24 08:21:42.031311 [ 9 ] <Debug> MemoryTracker: Peak memory usage: 76.81 MiB.
2018.05.24 08:21:42.031344 [ 4 ] <Debug> MemoryTracker: Peak memory usage: 155.58 MiB.
2018.05.24 08:21:42.031364 [ 16 ] <Debug> MemoryTracker: Peak memory usage: 65.25 MiB.
2018.05.24 08:21:42.031411 [ 14 ] <Debug> MemoryTracker: Peak memory usage: 60.19 MiB.
2018.05.24 08:21:42.031427 [ 7 ] <Debug> MemoryTracker: Peak memory usage: 54.28 MiB.
2018.05.24 08:21:42.037312 [ 1 ] <Debug> Application: Destroyed global context.
2018.05.24 08:21:42.056503 [ 1 ] <Error> Application: File not found: /data2/clickhouse/data/db/logs_shard/tmp_delete_032e3e9464547a440f0194235c9eeb94_4268213_4268213_0/datacenter.bin
2018.05.24 08:21:42.056555 [ 1 ] <Information> Application: shutting down
2018.05.24 08:21:42.056566 [ 1 ] <Debug> Application: Uninitializing subsystem: Logging Subsystem
2018.05.24 08:21:42.056698 [ 20 ] <Information> BaseDaemon: Stop SignalListener thread
2018.05.24 08:21:42.621160 [ 1 ] <Information> : Starting daemon with revision 54382
2018.05.24 08:21:42.624152 [ 1 ] <Information> Application: starting up
```

**Comment by kmatt** [comment]
I'm seeing something similar. The startup time before a node comes online is a problem. Any relation to #1758 ?

**Comment by ztlpn** [comment]
Millions of old parts is definitely not normal. During normal operation there should be about `old_parts_lifetime * <rate of inserts>` old parts. With the default value of old_parts_lifetime = 480 and recommended rate of insert ~ 1 per second we get ~1000 old parts.

Possible reasons that could cause that many parts to appear:
* High rate of INSERTs. Should be around once per second.
* Exotic partitioning. I've noticed that you use custom partitioning. How many partitions have you got?
* In case of Replicated tables, problems with deleting old parts info from ZooKeeper. Try searching for the `There is a problem with deleting part` string in logs.

**Comment by alexey-milovidov** [comment]
@dylanzr This issue is fixed in #4369
